### PR TITLE
Catch exceptions to show message before abort

### LINKED ISSUE #4028: Bug in particle parallelization
@egpuckett and I were working on creating a new particle interpolator, and ran into this error. The error also occurs with the cell average interpolator (although at a diffent timestep), which makes me think that either this vector is not getting safely operated on or that my parameter file is problematic. So far at least we have only encountered this error in 3D.
[The .prm used](https://drive.google.com/file/d/1ZfVVYAClYtLp_L0yk8-Ll9wnVQOMD2hB/view?usp=sharing)
```
--------------------------------------------------------
An error occurred in line <626> of file </mnt/data/Libraries/Ubuntu/dealii-9.2-nopetsc/tmp/unpack/deal.II-v9.2.0/source/lac/trilinos_vector.cc> in function
    void dealii::TrilinosWrappers::MPI::Vector::compress(dealii::VectorOperation::values)
The violated condition was: 
    result.max - result.min < 1e-5
Additional information: 
    Not all processors agree whether the last operation on this vector was an addition or a set operation. This will prevent the compress() operation from succeeding.

Stacktrace:
-----------
#0  /mnt/data/Libraries/Ubuntu/dealii-9.2-nopetsc/deal.II-v9.2.0/lib/libdeal_II.g.so.9.2.0: dealii::TrilinosWrappers::MPI::Vector::compress(dealii::VectorOperation::values)
#1  aspect: dealii::BlockVectorBase<dealii::TrilinosWrappers::MPI::Vector>::compress(dealii::VectorOperation::values)
#2  aspect: aspect::Simulator<3>::interpolate_particle_properties(aspect::Simulator<3>::AdvectionField const&)
#3  aspect: aspect::Simulator<3>::assemble_and_solve_composition(bool, std::vector<double, std::allocator<double> >*)
#4  aspect: aspect::Simulator<3>::solve_single_advection_single_stokes()
#5  aspect: aspect::Simulator<3>::solve_timestep()
#6  aspect: aspect::Simulator<3>::run()
#7  aspect: void run_simulator<3>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, bool, bool, bool)
#8  aspect: main
--------------------------------------------------------
```

### PR DESCRIPTION
**Pull request by gassmoeller**
This fixes an issue with weird MPI error messages we saw in our group, and should also fix #4028. Catching the exception and showing it's error message is sufficient to illustrate the real problem instead of the fake errors that result from MPI calls in destructors of objects further up in the callstack. I necessary we could also additionally call MPI_Abort, but at least for my problems this worked nicely already. @sac-bsa: Could you give this branch a try and see if your errors in #4028 look better now?

### CODE REVIEW

**Comment by sac-bsa** [comment]
I just tried it out. This works to ensure that the correct error is printed. Thank you!

File: source/particle/property/interface.cc [hunk 1]
```diff
@@ -393,10 +393,21 @@ namespace aspect
...
+                  try
+                    {
+                      interpolated_properties = interpolator.properties_at_points(particle_handler,
+                                                                                  std::vector<Point<dim> > (1,particle_location),
+                                                                                  ComponentMask(property_information.n_components(),true),
+                                                                                  found_cell);
+                    }
+                  catch (const std::exception &e)
+                    {
+                      std::cerr << e.what() << '\n';
```
Comment (tjhei) [inline source/particle/property/interface.cc:407]: I think it makes sense to use std::endl instead, to make sure output is flushed.

**Comment by tjhei** [comment]
Do you happen to remember which destructor is doing communication?

**Comment by gassmoeller** [comment]
Had to look it up, but it is `TimerOutput::Scope::~Scope()` in `deal.II/source/base/timer.cc:1045`. The MPI call is an MPI_Allreduce called from `dealii::Timer::stop`. I attach a stacktrace below:
```
#0  0x00007f646800b339 in ?? ()
   from /usr/lib/x86_64-linux-gnu/openmpi/lib/openmpi3/mca_btl_vader.so
#1  0x00007f64ad67a854 in opal_progress ()
   from /lib/x86_64-linux-gnu/libopen-pal.so.40
#2  0x00007f64ad681315 in ompi_sync_wait_mt ()
   from /lib/x86_64-linux-gnu/libopen-pal.so.40
#3  0x00007f64b0cb39f8 in ompi_request_default_wait ()
   from /lib/x86_64-linux-gnu/libmpi.so.40
#4  0x00007f64b0d0d83c in ompi_coll_base_sendrecv_actual ()
   from /lib/x86_64-linux-gnu/libmpi.so.40
#5  0x00007f64b0d0e042 in ompi_coll_base_allreduce_intra_recursivedoubling ()
   from /lib/x86_64-linux-gnu/libmpi.so.40
#6  0x00007f64b0cc9411 in PMPI_Allreduce ()
   from /lib/x86_64-linux-gnu/libmpi.so.40
#7  0x00007f64c5480ec8 in dealii::Utilities::MPI::internal::all_reduce<unsigned long> (mpi_op=@0x7fff7be8a298: 0x7f64b0d878a0 <ompi_mpi_op_min>, values=..., 
    mpi_communicator=@0x55d814284c78: 0x55d8110a2c90, output=...)
    at ../include/deal.II/base/mpi.templates.h:219
#8  0x00007f64c547fe45 in dealii::Utilities::MPI::min<unsigned long> (
    t=@0x7fff7be8a370: 1, mpi_communicator=@0x55d814284c78: 0x55d8110a2c90)
    at ../include/deal.II/base/mpi.templates.h:434
#9  0x00007f64c54705e4 in dealii::Utilities::MPI::min_max_avg (my_values=..., 
    result=..., mpi_communicator=@0x55d814284c78: 0x55d8110a2c90)
--Type <RET> for more, q to quit, c to continue without paging--
   ase/mpi.cc:682
#10 0x00007f64c546e83e in dealii::Utilities::MPI::min_max_avg (
    my_value=0.092740371000000002, mpi_communicator=@0x55d814284c78: 0x55d8110a2c90)
    at ../source/base/mpi.cc:94
#11 0x00007f64c5955a56 in dealii::Timer::stop (this=0x55d814284c40)
    at ../source/base/timer.cc:208
#12 0x00007f64c5956cc3 in dealii::TimerOutput::leave_subsection (this=0x7fff7be8ce38, 
    section_name="Particles: Interpolate") at ../source/base/timer.cc:467
#13 0x00007f64c595f0ab in dealii::TimerOutput::Scope::stop (this=0x7fff7be8ae10)
    at ../include/deal.II/base/timer.h:994
#14 0x00007f64c595bbec in dealii::TimerOutput::Scope::~Scope (this=0x7fff7be8ae10, 
    __in_chrg=<optimized out>) at ../source/base/timer.cc:1049
#15 0x000055d80dc49df7 in aspect::Simulator<2>::interpolate_particle_properties (
    this=0x7fff7be8c0d0, advection_field=...)
    at /home/gassmoel/software/aspect/source/simulator/initial_conditions.cc:198
#16 0x000055d80d8ed530 in aspect::Simulator<2>::assemble_and_solve_composition (
    this=0x7fff7be8c0d0, compute_initial_residual=false, initial_residual=0x0)
    at /home/gassmoel/software/aspect/source/simulator/solver_schemes.cc:262
#17 0x000055d80d8ee1f7 in aspect::Simulator<2>::solve_single_advection_single_stokes (
    this=0x7fff7be8c0d0)
    at /home/gassmoel/software/aspect/source/simulator/solver_schemes.cc:724
#18 0x000055d80d6dc738 in aspect::Simulator<2>::solve_timestep (this=0x7fff7be8c0d0)
    at /home/gassmoel/software/aspect/source/simulator/core.cc:1772
#19 0x000055d80d6d9733 in aspect::Simulator<2>::run (this=0x7fff7be8c0d0)
    at /home/gassmoel/software/aspect/source/simulator/core.cc:1952
#20 0x000055d80e493401 in run_simulator<2> (
    raw_input_as_string="# A setup that is simialr to the Tackley 2011 paper\n\n\nset Dimension", ' ' <repeats 30 times>, "= 2\nset Use years in output instead of seconds = true\nset End time", ' ' <repeats 31 times>, "= 4e8\n"..., 
    input_as_string="# A setup that is simialr to the Tackley 2011 paper\n\n\nset Dimension", ' ' <repeats 30 times>, "= 2\nset Use years in output instead of seconds = true\nset End time", ' ' <repeats 31 times>, "= 4e8\n"..., output_xml=false, output_plugin_graph=false, 
    validate_only=false) at /home/gassmoel/software/aspect/source/main.cc:623
#21 0x000055d80e44363b in main (argc=2, argv=0x7fff7be8f908)
    at /home/gassmoel/software/aspect/source/main.cc:817
```

**Comment by gassmoeller** [comment]
I addressed the comments. I also have an experimental fix for the TimerOutput::Scope destructor. Will give that a try and open as PR against deal.II. But that is independent of this fix.

File: source/simulator/initial_conditions.cc [hunk 2]
```diff
@@ -270,11 +270,21 @@ namespace aspect
...
+              particle_properties =
+                particle_interpolator->properties_at_points(particle_postprocessor.get_particle_world().get_particle_handler(),
+                                                            quadrature_points,
+                                                            property_mask,
+                                                            cell);
+            }
+          catch (const std::exception &e)
+            {
+              std::cerr << e.what() << std::endl;
+              throw e;
```
Comment (bangerth) [inline source/simulator/initial_conditions.cc:286]: This means that you will print the same error message twice. How about throwing `ExcQuiet` instead?

In some sense, what you're doing here is just adding debug output. In reality, if I understand correctly, you already found what the real problem is (namely calling `properties_at_point()` for invalid points). Do you even still need this patch then?

**Comment by bangerth** [comment]
@gassmoeller Ping?

**Comment by gassmoeller** [comment]
Sorry, yes, I still need this. The problem is that whether this exception triggers depends on the model setup, and that without this change the model will crash with completely unrelated and misleading error messages. I have updated the output to use the same code we use for the postprocessors in postprocess/interface.cc. It would be good if this could go into 2.3 (since we already had at least 2 people encountering this problem).

**Review by tjhei — APPROVED** [review]
Looks good to me. Ready to merge, Rene?

**Comment by gassmoeller** [comment]
Yes, ready from my side.

**Pull request merged**